spring-batch-test/src/main/java/org/springframework/batch/test/JobLauncherTestUtils.java

```diff
@@ -155,12 +156,12 @@ public JobExecution launchJob(JobParameters jobParameters) throws Exception {
 	}
 
 	/**
-	 * @return a new JobParameters object containing only a parameter for the
-	 * current timestamp, to ensure that the job instance will be unique.
+	 * @return a new JobParameters object containing only a UUID parameter,
+	 * to ensure that the job instance will be unique.
 	 */
 	public JobParameters getUniqueJobParameters() {
 		Map<String, JobParameter> parameters = new HashMap<>();
-		parameters.put("random", new JobParameter((long) (Math.random() * JOB_PARAMETER_MAXIMUM)));
+		parameters.put("random", new JobParameter(UUID.randomUUID().toString()));

```

Comment:
While I understand that iterating over 1000000 numbers randomly may not provide enough entropy to prevent duplicates, I feel like we should be able to figure out a way to address this in a non-breaking change.  Even something like taking the current time (long) and adding a random number to it should be good enough...  I'm thinking of 

`parameters.put("random", new JobParameter(new Date().getTime() + (long) (Math.random() * JOB_PARAMETER_MAXIMUM)));`

While not super elegant, it prevents the potential of breaking thousands of tests in the community, most of which have not been impacted by this issue.

Reply:
I understand the need to minimize breaking changes.  I like your suggestion. In addition to the suggest change I would push for a new method (something like **getRealUniqueJobParameters** but with a better name) for users who are using this method in their test framework and want a stronger assurance that the method returns unique parameters.

So basically I'm proposing:
1. Let's add a non-breaking change using your suggestion to add more entropy in the existing method
2. Let's create a new method that uses UUID (or something else) to make a stronger guarantee of actual uniqueness.
3. There should be a decision on which of these methods JobLauncherTestUtils.launchJob should call. 

To give context, we were getting almost daily job failures on our integration tests because we were using `jobLauncherTestUtils.launchJob` which calls this. We ended up writing our own version of this test utils which uses UUID generation to generate the job parameters.

Reply:
I understand the need to minimize breaking changes. I like your suggestion. In addition to the suggest change I would push for a new method (something like getRealUniqueJobParameters but with a better name) for users who are using this method in their test framework and want a stronger assurance that the method returns unique parameters.

So basically I'm proposing:

Let's add a non-breaking change using your suggestion to add more entropy in the existing method
Let's create a new method that uses UUID (or something else) to make a stronger guarantee of actual uniqueness.
There should be a decision on which of these methods JobLauncherTestUtils.launchJob should call.
To give context, we were getting almost daily job failures on our integration tests because we were using jobLauncherTestUtils.launchJob which calls this. We ended up writing our own version of this test utils which uses UUID generation to generate the job parameters.

Reply:
I like the idea of a new method.  We could even depricate the current one in favor of it.  I also agree that we'd need a better name than `getRealUniqueJobParameters` ;)  I'd also vote to leave the `JobLauncherTestUtils.launchJob` method alone until the old method is actually removed.

Reply:
I see no need to add a new method, we can fix this in a non-breaking way. However, I would not roll my own random number generation logic but delegate this to a class provided by the JDK. `UUID.randomUUID().timestamp()` would have worked (from its Javadoc its results in a "60 bit timestamp measured in 100-nanosecond units since midnight, October 15, 1582 UTC"), but unfortunately Java does not support the generation of time based UUIDs (`UUID.randomUUID()` generates a UUID of type 4 which is not time based, so this fails with `java.lang.UnsupportedOperationException: Not a time-based UUID`).

The other option is to use a `java.security.SecureRandom`:

```diff
- parameters.put("random", new JobParameter((long) (Math.random() * JOB_PARAMETER_MAXIMUM)));
+ parameters.put("random", new JobParameter(new SecureRandom().nextLong()));
```

`SecureRandom` is a cryptographically strong random number generator which addresses most of the limitations of pseudo random generators like `java.util.Random` used behind the scene by `Match.random()` in the current code. `SecureRandom` is actually what's used by the UUID class to generate random UUIDs.

There are a couple of advantages of using this approach:

- It is not a breaking change (the type of the `random` job parameter is still `long`)
- It delegates random number generation to class provided by the JDK

BTW, the test added in this PR passes with this approach even if I increase the upper bound to 1 million elements (instead of 10k).

Reply:
I'm hesitating a bit here because we are talking about a domain that I have no expertise in (how to generate UUID like numbers in a long). To take a step back let's clarify the requirements:

1.  **To maintain backwards compatability** - this means we need to have a random JobParameters with type long to fit past behavior.
2. **To have a low likelihood of a repeat** - A repeated number could result in a job failure of and breaks the method's contract of generating a unique"job parameter. I believe my original proposed solution used UUID's which have around 1 in 50 billion chance of collision, but unfortunately is not a practical solution due to it's data type. We probably don't need that strong of a collision guarantee but we definitely something much better than the current random number.

@benas - how does your proposed solution compare with just using something like UUID.randomUUID().getMostSignificantBits()

Again, I don't have any expertise in this area but the overall objective in my mind is to "reduce the probability of collision of a generated long value" so I just want to take a step back and see if there are any more appropriate JDK methods.